### PR TITLE
Fix renaming symlinks on Windows

### PR DESCRIPTION
**Pull request by ChrisDenton**
Previously we only detected mount points and not other types of links when determining reparse point behaviour.

Also added some tests to avoid this regressing again in the future.

### CODE REVIEW

File: library/std/src/fs/tests.rs [hunk 1]
```diff
@@ -1953,3 +1953,24 @@ fn test_rename_directory_to_non_empty_directory() {
...
+#[test]
+#[cfg(windows)]
+fn test_rename_junction() {
+    let tmpdir = tmpdir();
+    let original = tmpdir.join("original");
+    let dest = tmpdir.join("dest");
+
+    junction_point("does not exist", &original).unwrap();
+    fs::rename(&original, &dest).unwrap();
+}
```
Comment (tgross35) [inline library/std/src/fs/tests.rs:1976]: Should these tests include a check that the symlink is accessible via `dest`?

The changes lgtm

Comment (ChrisDenton) [reply library/std/src/fs/tests.rs:1976]: Oh wait, maybe I don't have to test it literally works just that `fs::read_link` returns that it's indeed a link to the file we expected.

File: library/std/src/fs/tests.rs [hunk 2]
```diff
@@ -1953,3 +1953,32 @@ fn test_rename_directory_to_non_empty_directory() {
...
+fn test_rename_junction() {
+    let tmpdir = tmpdir();
+    let original = tmpdir.join("original");
+    let dest = tmpdir.join("dest");
+    let not_exist = Path::new("does not exist");
+
+    junction_point(&not_exist, &original).unwrap();
+    fs::rename(&original, &dest).unwrap();
+
+    // Make sure that copying `original`` to `dest` preserves the junction point.
```
Comment (tgross35) [inline library/std/src/fs/tests.rs:1981]: Nit

```suggestion
    // Make sure that copying `original` to `dest` preserves the junction point.
```

File: library/std/src/fs/tests.rs (hunk 1 above)
Comment (ChrisDenton) [reply library/std/src/fs/tests.rs]: Ok, I did the above.

**Review by tgross35 — APPROVED** [review]
Great idea with `read_link`, r=me with the nit and a squash.

**Pull request merged**